opm/parser/eclipse/Parser/Parser.cpp

```diff
@@ -42,60 +42,194 @@
+            file( boost::filesystem::path p, const std::string& in ) :
+                path( p ), input( in )
+            {}
+
+            file( const std::string& in ) : path( "" ), input( in ) {}
+
+            boost::filesystem::path path;
+            string_view input;
+            size_t lineNR = 0;
+        };
+
         const ParseContext& parseContext;
         Deck * deck;
-        boost::filesystem::path dataFile;
+
         boost::filesystem::path rootPath;
-        /* the path map is shared between all state instances, as it has global
-         * resolution, i.e. PATHS specified in an included file can be used
-         * from the file that included it.
-         */
-        std::shared_ptr< std::map<std::string, std::string> > pathMap;
-        size_t lineNR;
-        std::shared_ptr<std::istream> inputstream;
+        std::map< std::string, std::string > pathMap;
+        std::list< std::string > input_strings;
+        std::list< file > input_files;
+        std::vector< file* > input_stack;
         RawKeywordPtr rawKeyword;
-        std::string nextKeyword;
+        string_view nextKeyword = emptystr;
 
+        boost::filesystem::path& current_path() {
+            return this->input_stack.back()->path;
+        }
 
-        ParserState(const ParserState& parent)
-            : parseContext( parent.parseContext )
-        {
-            deck = parent.deck;
-            pathMap = parent.pathMap;
-            rootPath = parent.rootPath;
+        const boost::filesystem::path& current_path() const {
+            return this->input_stack.back()->path;
         }
 
-        ParserState(const ParseContext& __parseContext)
-            : parseContext( __parseContext ),
-              deck( new Deck() ),
-              pathMap( std::make_shared< std::map< std::string, std::string > >() ),
-              lineNR( 0 )
-        {}
+        size_t& line() {
+            return this->input_stack.back()->lineNR;
+        }
 
-        std::shared_ptr<ParserState> includeState(boost::filesystem::path& filename) {
-            std::shared_ptr<ParserState> childState = std::make_shared<ParserState>( *this );
-            childState->openFile( filename );
-            return childState;
+        const size_t& line() const {
+            return this->input_stack.back()->lineNR;
         }
 
+        string_view& input() {
+            return this->input_stack.back()->input;
+        }
 
-        void openString(const std::string& input) {
-            dataFile = "";
-            inputstream.reset(new std::istringstream(input));
+        void pop() {
+            this->input_stack.pop_back();
         }
 
+        ParserState(const ParseContext& __parseContext)
+            : parseContext( __parseContext ),
+              deck( new Deck() )
+        {}
 
-        void openStream(std::shared_ptr<std::istream> istream) {
-            dataFile = "";
-            inputstream = istream;
+        void openString(const std::string& input ) {
```

Comment:
Since this is not streaming anymore - can we have `loadString` and also `loadFile`?